node/src/components/consensus/protocols/zug/des_testing.rs

```diff
@@ -0,0 +1,1234 @@
+                                    Content::Echo(HashWrapper(hash.0.wrapping_add(1))),
+                                    signed_msg.validator_idx,
+                                    &TestSecret(signed_msg.validator_idx.0.into()),
+                                );
+                                vec![
+                                    ZugMessage::GossipMessage(EraMessage::Zug(Box::new(
+                                        ZugProtocolMessage::Signed(conflicting_message),
+                                    ))),
+                                    msg,
+                                ]
+                            }
+                            Content::Vote(vote) => {
+                                let conflicting_message = SignedMessage::sign_new(
+                                    signed_msg.round_id,
+                                    signed_msg.instance_id,
+                                    Content::Vote(!vote),
+                                    signed_msg.validator_idx,
+                                    &TestSecret(signed_msg.validator_idx.0.into()),
+                                );
+                                vec![
+                                    ZugMessage::GossipMessage(EraMessage::Zug(Box::new(
+                                        ZugProtocolMessage::Signed(conflicting_message),
+                                    ))),
+                                    msg,
+                                ]
+                            }
+                        }
+                    }
+                    _ => vec![msg],
+                },
+                _ => vec![msg],
+            },
+        }
+    }
+}
+
+type ZugNode = Node<ConsensusValue, ZugMessage, ZugValidator>;
+
+type ZugNet = VirtualNet<ConsensusValue, ZugMessage, ZugValidator>;
+
+struct ZugTestHarness<DS>
+where
+    DS: DeliveryStrategy,
+{
+    virtual_net: ZugNet,
+    /// Consensus values to be proposed.
+    /// Order of values in the vector defines the order in which they will be proposed.
+    consensus_values: VecDeque<ConsensusValue>,
+    /// A strategy to pseudo randomly change the message delivery times.
+    delivery_time_strategy: DS,
+    /// Distribution of delivery times.
+    delivery_time_distribution: Distribution,
+    /// Mapping of validator IDs to node IDs
+    vid_to_node_id: HashMap<ValidatorId, NodeId>,
+    /// Mapping of node IDs to validator IDs
+    node_id_to_vid: HashMap<NodeId, ValidatorId>,
+}
+
+type TestResult<T> = Result<T, TestRunError>;
+
+impl<DS> ZugTestHarness<DS>
+where
+    DS: DeliveryStrategy,
+{
+    /// Advance the test by one message.
+    ///
+    /// Pops one message from the message queue (if there are any)
+    /// and pass it to the recipient validator for execution.
+    /// Messages returned from the execution are scheduled for later delivery.
+    pub(crate) fn crank(&mut self, rng: &mut NodeRng) -> TestResult<()> {
```

Comment:
While they're implemented elsewhere, Zug info messages could be made more informative and put into some sort of standard format, to facilitate quantitative analysis of multiple runs. E.g., an info message surfaced by the protocol should always contain the originating validator id.

```
2023-03-14T19:34:58.374175Z   ERROR [c:c:c:p:zug :700] we are faulty; deactivating; our_idx=1
2023-03-14T19:34:58.374616Z INFO  [c:c:c:p:zug :1231] round is now skippable;   round_id=1
2023-03-14T19:34:58.374687Z INFO  [c:c:c:p:zug :1545] started a new round;   round_id=2; leader=11
2023-03-14T19:34:58.375207Z INFO  [c:c:c:p:zug :1231] round is now skippable;   round_id=1
2023-03-14T19:34:58.375277Z INFO  [c:c:c:p:zug :1545] started a new round;   round_id=2; leader=11
```

For example, who is faulty above? Who are the (presumably 2) validators who just detected the start of a new round? Ids would help with reconstructing the evolution of the protocol state over time.

Reply:
I've added `our_idx` field to all log messages in Zug in ba5bf02. There are a few drawbacks, though:
- It's the validator _index_, not the public key - which means it might be a bit harder to make the connection between a validator and proposer's key in a block, for example,
- It's only logged if the validator is _active_, that is, when it's bonded in the given era and is creating messages.
Zug isn't aware of its validator ID otherwise.
That should still be an improvement, though maybe it's not quite perfect.

Reply:
Looks great from just reading the code, at least. I guess it's time to merge.